File: src/main/java/cn/huangsy/DecodeString.java
package cn.huangsy;

/**
 * @author huangsy
 * @date 2020/5/28 8:40
 */
public class DecodeString {

    public static final char start = '[';
    public static final char end = ']';
    public int i = 0;

    public String decodeString(String s) {
        return decode(s.toCharArray());
    }

    private String decode(char[] chars) {
        String x = "";
        while (true && i < chars.length) {
            if (chars[i] <= 47 || chars[i] >= 58) {
                if (chars[i] == start) {
                    int n = getNumber(chars, i-1);
                    i ++;
                    String m = decode(chars);
                    for (int j = 0; j < n; j++) {
                        x = x.concat(m);
                    }
                } else if (chars[i] == end) {
                    break;
                } else {
                    x = x.concat(Character.toString(chars[i]));
                }
            }
            i++;
        }
        return x;
    }

    private int getNumber(char[] chars, int i) {
        int n = 0;
        int w = 1;
        while (i >= 0 && chars[i] > 47 && chars[i] < 58) {
            n += (chars[i] - 48 )* w;
            i--;
            w *= 10;
        }
        return n;
    }

    public static void main(String[] args) {
        /**
         * s = "3[a]2[bc]", 返回 "aaabcbc".
         * s = "3[a2[c]]", 返回 "accaccacc".
         * s = "2[abc]3[cd]ef", 返回 "abcabccdcdcdef".
         */
        DecodeString decodeString = new DecodeString();
//        System.out.println(decodeString.decodeString("3[a]2[bc]"));
//        System.out.println(decodeString.decodeString("3[a2[c]]"));
        System.out.println(decodeString.decodeString("2[abc]3[cd]ef"));
    }
}

File: src/main/java/cn/leetcode/394/DecodeString.kt
package cn.leetcode.`394`

import cn.huangsy.ValidPalindrome

/**
 * @author huangsy
 * @date 2020/5/28 10:08
 */
class DecodeString {
    val start = '['
    val end = ']'
    var i = 0

    fun decodeString(s: String): String? {
        return decode(s.toCharArray())
    }

    private fun decode(chars: CharArray): String {
        var x = ""
        while (true && i < chars.size) {
            if (chars[i] <= 47.toChar() || chars[i] >= 58.toChar()) {
                if (chars[i] == start) {
                    val n = getNumber(chars, i - 1)
                    i++
                    val m = decode(chars)
                    for (j in 0 until n) {
                        x = x + m
                    }
                } else if (chars[i] == end) {
                    break
                } else {
                    x = x + Character.toString(chars[i])
                }
            }
            i++
        }
        return x
    }

    private fun getNumber(chars: CharArray, i: Int): Int {
        var i = i
        var n = 0
        var w:Int = 1
        while (i >= 0 && chars[i] > 47.toChar() && chars[i] < 58.toChar()) {
            n += ((chars[i] - 48).toInt() * w)
            i--
            w *= 10
        }
        return n
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val decodeString = DecodeString()
            println(decodeString.decodeString("2[abc]3[cd]ef"))
        }
    }
}
